SwiftMonkey/MonkeyXCTest.swift

```diff
@@ -40,4 +40,80 @@ extension Monkey {
             }
         }
     }
+    
+    
+    public func addDefaultXCTestPublicActions() {
+        addXCTestPublicTapAction(weight: 25)
+        addXCTestPublicLongPressAction(weight: 1)
+        addXCTestPublicDragAction(weight: 1)
+    }
+    
+    /**
+     Add an action that generates a tap, with a possibility for
+     multiple taps with multiple fingers, using the private
+     XCTest API.
+     
+     - parameter weight: The relative probability of this
+     event being generated. Can be any value larger than
+     zero. Probabilities will be normalised to the sum
+     of all relative probabilities.
+     - parameter multipleTouchProbability: Probability that
+     the tap event will use multiple fingers. Between 0 and 1.
+     */
+    public func addXCTestPublicTapAction(weight: Double,
+                                   multipleTouchProbability: Double = 0.05) {
+        addAction(weight: weight) { [weak self] in
+            let locations: [CGPoint]
+            if self!.r.randomDouble() < multipleTouchProbability {
+                let numberOfTouches = Int(self!.r.randomUInt32() % 3) + 2
+                let rect = self!.randomRect()
+                locations = (1...numberOfTouches).map { _ in
+                    self!.randomPoint(inRect: rect)
+                }
+            } else {
+                locations = [ self!.randomPoint() ]
+            }
+            let app = XCUIApplication()
+            let coordinate = app.coordinate(withNormalizedOffset: CGVector(dx: locations[0].x/(app.frame.maxX/2), dy: locations[0].y/(app.frame.maxY/2)))
+            coordinate.tap()
+        }
+    }
```

Comment:
Hello and thanks for the feedback! Your implementation does not really utilize multiple touches, as this API is not available on the `XCUICoordinate` object. Also I noticed that touches have tendency to cluster on the edges of the screen:

<table>
<tr>
<td>
<img width="400px" src="https://user-images.githubusercontent.com/1027098/66109757-aa4cf900-e5c5-11e9-99f5-cda289f6bb83.png"/>
</td>
</tr>
</table>

I am also concerned by `let app = XCUIApplication()` inside the action. According to #70 it might cause memory bloat. Here is how I would change the code:

```suggestion
    /// Add an action that generates a tap, with a possibility for double taps, using the public XCTest API.
    /// - Parameter app: The application proxy.
    /// - Parameter weight: The relative probability of this event being generated. Can be any value larger than zero. Probabilities
    /// will be normalised to the sum of all relative probabilities.
    /// - Parameter doubleTapProbability: Probability that a double tap event is used. Between 0 and 1.
    public func addXCTestPublicTapAction(app: XCUIApplication,
                                         weight: Double,
                                         doubleTapProbability: Double = 0.05) {
        addAction(weight: weight) { [unowned self] in
            let doubleTap = self.r.randomDouble() < doubleTapProbability
            let coordinate = app.coordinate(withNormalizedOffset: .zero).withOffset(self.randomOffset())
            
            if doubleTap {
                coordinate.doubleTap()
            } else {
                coordinate.tap()
            }
        }
    }
```

Notice that I've replaced `multipleTouchProbability` with `doubleTapProbability` as this is supported directly by the `XCUICoordinate` class. I haven't tested though whether tapping immediately multiple coordinates would generate a multi-touch event. I would also add the `app` parameter to the `addDefaultXCTestPublicActions` method, e.g:

```swift
public func addDefaultXCTestPublicActions(app: XCUIApplication) {
    addXCTestPublicTapAction(app: app, weight: 25)
    addXCTestPublicLongPressAction(app: app, weight: 1)
    addXCTestPublicDragAction(app: app, weight: 1)
}
```

What do you think @a455455b ?

Reply:
I forgot to attach the code for `randomVector()` method (though it's very straightforward):

```swift
/// Generates a random `CGVector` inside the frame of the app.
public func randomOffset() -> CGVector {
    let point = randomPoint()
    return CGVector(dx: point.x, dy: point.y)
}
```

Reply:
Very good, it gave me a lot of inspiration.

Reply:
Need me to modify and resubmit?

Reply:
Yes, ideally you could try adapting other methods that you've added (add the `app` parameter, compute coordinates properly, etc.).

If you wish I can also do it myself by providing code suggestions.

Reply:
hi，I submitted a fix.